src/analysis/convex/extreme.lean

```diff
@@ -175,13 +168,58 @@ subset.antisymm (λ x hx, ⟨hx.1, hAB.extreme_points_subset_extreme_points hx
 end has_smul
 
 section ordered_semiring
-variables {𝕜} [ordered_semiring 𝕜] [add_comm_group E] [module 𝕜 E] {A B : set E} {x : E}
+variables [ordered_semiring 𝕜] [add_comm_group E] [add_comm_group F] [Π i, add_comm_group (π i)]
+  [module 𝕜 E] [module 𝕜 F] [Π i, module 𝕜 (π i)] {A B : set E} {x : E}
 
 lemma is_extreme.convex_diff (hA : convex 𝕜 A) (hAB : is_extreme 𝕜 A B) :
   convex 𝕜 (A \ B) :=
 convex_iff_open_segment_subset.2 (λ x₁ ⟨hx₁A, hx₁B⟩ x₂ ⟨hx₂A, hx₂B⟩ x hx,
     ⟨hA.open_segment_subset hx₁A hx₂A hx, λ hxB, hx₁B (hAB.2 hx₁A hx₂A hxB hx).1⟩)
 
+@[simp] lemma extreme_points_prod (s : set E) (t : set F) :
+  (s ×ˢ t).extreme_points 𝕜 = s.extreme_points 𝕜 ×ˢ t.extreme_points 𝕜 :=
+begin
+  ext,
+  refine (and_congr_right $ λ hx, ⟨λ h, _, λ h, _⟩).trans (and_and_and_comm _ _ _ _),
+  split,
+  { rintro x₁ hx₁ x₂ hx₂ hx_fst,
+    refine (h (mk_mem_prod hx₁ hx.2) (mk_mem_prod hx₂ hx.2) _).imp
+      (congr_arg prod.fst) (congr_arg prod.fst),
+    rw ←prod.image_mk_open_segment_left,
+    exact ⟨_, hx_fst, prod.mk.eta⟩ },
+  { rintro x₁ hx₁ x₂ hx₂ hx_snd,
+    refine (h (mk_mem_prod hx.1 hx₁) (mk_mem_prod hx.1 hx₂) _).imp
+      (congr_arg prod.snd) (congr_arg prod.snd),
+    rw ←prod.image_mk_open_segment_right,
+    exact ⟨_, hx_snd, prod.mk.eta⟩ },
+  { rintro x₁ hx₁ x₂ hx₂ ⟨a, b, ha, hb, hab, hx'⟩,
+    simp_rw prod.ext_iff,
+    exact (and_and_and_comm _ _ _ _).1
+     ⟨h.1 hx₁.1 hx₂.1 ⟨a, b, ha, hb, hab, congr_arg prod.fst hx'⟩,
+      h.2 hx₁.2 hx₂.2 ⟨a, b, ha, hb, hab, congr_arg prod.snd hx'⟩⟩ }
+end
+
+@[simp] lemma extreme_points_pi (s : Π i, set (π i)) :
+  (univ.pi s).extreme_points 𝕜 = univ.pi (λ i, (s i).extreme_points 𝕜) :=
+begin
+  ext,
+  simp only [mem_extreme_points, mem_pi, mem_univ, true_implies_iff, @forall_and_distrib ι],
+  refine and_congr_right (λ hx, ⟨λ h i, _, λ h, _⟩),
+  { rintro x₁ hx₁ x₂ hx₂ ⟨a, b, ha, hb, hab, hx'⟩,
+    refine (h (update x i x₁) _ (update x i x₂) _ _).imp (λ h₁, by rw [←h₁, update_same])
+      (λ h₂, by rw [←h₂, update_same]);
+    try { rintro j,
```

Comment:
How many goals are there at this `try`?

Reply:
Three, the first two of which are dispatched with the `try`.

Reply:
Can you use `rotate` to dispatch the other goals first, then `all_goals` instead? `try` is usually problematic since it's not clear which goals it was intended to apply to.

Reply:
Or alternatively
```suggestion
    iterate 2 { rintro j,
```